File: 3-longest-substring-without-repeating-characters/3-longest-substring-without-repeating-characters.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        track = dict()
        n = len(s)
        start  = 0
        res = 0
        for i in range(n):
            if s[i] in track and track[s[i]]>=start:
                start = track[s[i]] + 1
            else:                      
                res = max(res,i-start+1)
            track[s[i]] = i
            
        return res       

File: 3-longest-substring-without-repeating-characters/3-longest-substring-without-repeating-characters.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        if(s.length()<2){
            return s.length();
        }
        Map<Character,Integer> map = new HashMap<>();
        int start = 0;
        int res =0;
        for(int i =0;i<s.length();i++){
            if(map.containsKey(s.charAt(i)) && map.get(s.charAt(i))>=start){
                start = map.get(s.charAt(i))+1;   
            }
            else{
                res = Math.max(res,i-start+1);
                System.out.println(res);
            }
            
            map.put(s.charAt(i),i);
            
            
        }
        
        return res;
        
    }
}